libs/platform/ebpf_hash_table.c

```diff
@@ -3,52 +3,321 @@
+    ebpf_hash_bucket_header_t* volatile buckets[1];
 };
 
-// NOTE:
-// AVL tree gives a single struct containing both key and value.
-// Compare can be called with a partial struct only containing the key.
-// Do not access beyond map->ebpf_map_definition.key_size bytes.
-static RTL_GENERIC_COMPARE_RESULTS
-_ebpf_hash_map_compare(_In_ struct _RTL_AVL_TABLE* avl_table, _In_ void* first_struct, _In_ void* second_struct)
+/**
+ * @brief Perform a rotate left on a value.
+ *
+ * @param[in] value Value to be rotated.
+ * @param[in] count Count of bits to rotate.
+ * @return Rotated value.
+ */
+static inline unsigned long
+_ebpf_rol(uint32_t value, size_t count)
+{
+    return (value << count) | (value >> (32 - count));
+}
+
+// Ported from https://github.com/aappleby/smhasher
+// Quote from https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L2
+// "MurmurHash3 was written by Austin Appleby, and is placed in the public domain."
+
+/**
+ * @brief An implementation of the murmur3_32 hash function. This is a high
+ * performance non-cryptographic hash function.
+ *
+ * @param[in] key Pointer to key to hash.
+ * @param[in] length Length of key to hash.
+ * @param[in] seed Seed to randomize hash.
+ * @return Hash of key.
+ */
+unsigned long
+_ebpf_murmur3_32(_In_ const uint8_t* key, size_t length, uint32_t seed)
 {
-    ebpf_hash_table_t* table = (ebpf_hash_table_t*)avl_table;
+    uint32_t c1 = 0xcc9e2d51;
+    uint32_t c2 = 0x1b873593;
+    uint32_t r1 = 15;
+    uint32_t r2 = 13;
+    uint32_t m = 5;
+    uint32_t n = 0xe6546b64;
+    uint32_t hash = seed;
+
+    for (size_t index = 0; (length - index) > 3; index += 4) {
+        uint32_t k = *(uint32_t*)(key + index);
+        k *= c1;
+        k = _ebpf_rol(k, r1);
+        k *= c2;
+
+        hash ^= k;
+        hash = _ebpf_rol(hash, r2);
+        hash *= m;
+        hash += n;
+    }
+    unsigned long remainder = 0;
+    for (size_t index = length & (~3); index < length; index++) {
+        remainder <<= 8;
+        remainder |= key[index];
+    }
+    remainder *= c1;
+    remainder = _ebpf_rol(remainder, r1);
+    remainder *= c2;
+
+    hash ^= remainder;
+    hash ^= (uint32_t)length;
+    hash *= 0x85ebca6b;
+    hash ^= (hash >> r2);
+    hash *= 0xc2b2ae35;
+    hash ^= (hash >> 16);
+    return hash;
+}
 
-    if (table->compare_function) {
-        return (RTL_GENERIC_COMPARE_RESULTS)table->compare_function(
-            (const uint8_t*)first_struct, (const uint8_t*)second_struct);
+/**
+ * @brief Given two potentially non-comparable key values, extract the key and
+ * compare them.
+ *
+ * @param[in] hash_table Hash table the keys belong to.
+ * @param[in] key_a First key
+ * @param[in] key_b Second key
+ * @retval -1 if key_a < key_b
+ * @retval 0 if key_a == key_b
+ * @retval 1 if key_a > key_b
+ */
+static int
+_ebpf_hash_table_compare(ebpf_hash_table_t* hash_table, _In_ const uint8_t* key_a, _In_ const uint8_t* key_b)
```

Comment:
add SAL annotation to hash_table argument.  Is it `_Inout_` or should it be `_In_ const ebpf_hash_table_t*`?